1-Draft/RFC0022-RFC-Test-Connection-Cross-Platform.md

```diff
@@ -3,29 +3,167 @@ RFC: RFC0022
 
 ## Motivation
 
-The motivation for this RFC is as simple as the ping command itself; provide a Test-Connection cmdlet that can run on any platform that PowerShell can.  This means stripping out some of the WSMan specific remoting capabilities that have been added to allow for distributed execution and removing WMI based methods and classes in favor of pure C# code.
+PowerShell users need an cross-platform, available and power tool to troubleshoot and diagnose a network.
 
-This could also be back-ported into the Windows implementation, which while immediately removing the remoting capability would increase the performance of the cmdlet by not requiring WMI accessors and allow for tighter integration between multiple platforms that this cmdlet could run on.
+Currently users can use some native utilites such as `nmap`, `ping`, `traceroute`, `pathping`. We want to have a cmdlet that provides most of these capabilities.
 
-It would also serve as a framework for further enhancement, such as using SSHRemoting to enable the distributed approach to performing a distributed ping, while allowing for some built in governor mechanism to prevent abuse of the distributed nature.
+## Specification
 
-Lastly, and most importantly, while the current Test-Connection cmdlet in the Windows invocation of PowerShell is limited, it is used extensively to check for the existence and responsiveness of remote systems.  By not providing a non-Windows specific implementation that is as close in parameter set and return types then existing code must be re-written to overcome this limitation, or adoption of PowerShell on non-Windows platforms may be barred by the lack of this simple but widely used functionality.
+### Outputs
 
-## Specification
+- (empty yet)
 
-## Alternate Proposals and Considerations
+### Parameter Sets
+
+1. Ping - (1) resolve hostname address and ping IPv6 address or fallback to IPv4 address. (2) detect input address type anf ping IPv6 or IPv4 address.
+
+1. TraceRoute
+
+1. PathPing
+
+1. ConnectionByTCPPortName
+
+1. ConnectionByUDPPortName
+
+1. ConnectionByTCPPort
+
+1. ConnectionByUDPPort
+
+1. DetectionOfBlackHole - detect Black Hole Router. See [Diagnoses and treatment of black hole routers](https://support.microsoft.com/en-us/help/159211/diagnoses-and-treatment-of-black-hole-routers)
+
+1. DetectionOfMTUSize - detect MTU size. See [Recommended TCP/IP settings for WAN links with a MTU size of less than 576](https://support.microsoft.com/en-us/help/900926/recommended-tcp-ip-settings-for-wan-links-with-a-mtu-size-of-less-than)
+
+### Common Parameters
+
+1. InformationalLevel
+
+   - Quiet - return true/false or a single value of the response.
+   - Detailed - return an result object corresponding the parameterset.
+
+1. TimeOut - timeout for each request (not cmdlet).
+    [int]
+    ValidatePositive()
+
+1. ComputerName - hostname(s) or IPv4/IPv6 address(es).
```

Comment:
Perhaps support for CIDR blocks and wildcards?

```powershell
Test-Connection -ComputerName "10.0.0.0/24"
Test-Connection -ComputerName "10.0.0.*"
Test-Connection -ComputerName @(0..255 | %{"10.0.0.$_"})
```

Wildcards maybe not a great idea.. but CIDR is really helpful even for IPv6

Reply:
We may not even mention it here - as we add _parallelism_ this will be implied.

Reply:
what?

Reply:
Sequential ping of multiple addresses is too slow. We need to get support for parallel execution in PowerShell engine first. After that, the ping of address ranges will be the first feature to be implemented

Reply:
But you already support pinging an address range via `[string[]]`. Prior to execution, build the actual list of IP's to scan through iterations and expansion. Then either scan that list serially or concurrently.. at this point, maybe without concurrent processing, but the process would be the same when concurrency is available. 

Besides, this can be done concurrently now through several means. I don't think concurrency is predicated on the PowerShell implementation as the actual network tests are being in in .NET. 

Regardless of concurrency, support of network ranges could be implemented.

Reply:
We could await ping. But as it turned out this code has bugs and poorly tested. We will definitely do this but in the second stage - the current RFC does not forbid it.

Reply:
Just skip to the end game and port and wrap Nmap :)

Reply:
Just skip to the end game and port and wrap Nmap :)